mantle-build/src/dispatcher_gen.rs

```diff
@@ -112,40 +118,82 @@ fn generate_rpc_dispatcher(
             let mut service = <{service_ident}>::coalesce();
             let payload: RpcPayload =
                 mantle::reexports::serde_cbor::from_slice(&mantle::backend::input()).unwrap();
-            let output = match payload {{
-                {call_tree} // match arms return ABI-encoded vecs
+            let output: std::result::Result<Vec<u8>, {output_err_ty}> = match payload {{
+                {call_tree}
                 {default_fn_arm}
             }};
             <{service_ident}>::sunder(service);
             match output {{
                 Ok(output) => mantle::backend::ret(&output),
-                Err(err) => mantle::backend::err(&format!("{{:#?}}", err).into_bytes()),
+                Err(err_output) => {err_returner},
             }}
         }}
         }}"#,
-        rpc_payload_variants = rpc_payload_variants,
+        rpc_payload_variants = rpc_payload_variants.join(", "),
         service_ident = service_name.as_str().get(),
         call_tree = rpc_match_arms,
         default_fn_arm = default_fn_arm,
+        output_err_ty = output_err_ty,
+        err_returner = err_returner
     ) => parse_block)
 }
 
+fn gen_result_dispatch(name: Symbol, arg_name_csv: impl AsRef<str>) -> String {
+    format!(
+        r#"RpcPayload::{name}({args}) => match service.{name}(&ctx, {args}) {{
+            Ok(output) => Ok(mantle::reexports::serde_cbor::to_vec(&output).unwrap()),
+            Err(err) => Err(mantle::reexports::serde_cbor::to_vec(&err).unwrap()),
+        }}"#,
+        name = name,
+        args = arg_name_csv.as_ref(),
+    )
+}
+
+fn gen_dispatch(name: Symbol, arg_name_csv: impl AsRef<str>) -> String {
+    format!(
+        r#"RpcPayload::{name}({args}) => {{
+            Ok(mantle::reexports::serde_cbor::to_vec(&service.{name}(&ctx, {args})).unwrap())
+        }}"#,
+        name = name,
+        args = arg_name_csv.as_ref(),
+    )
+}
+
 fn generate_ctor_fn(service_name: Symbol, ctor: &MethodSig) -> P<Item> {
-    let ctor_arg_names = ctor.decl.inputs[1..]
-        .iter()
-        .map(|arg| pprust::pat_to_string(&arg.pat))
-        .collect::<Vec<_>>()
-        .join(", ");
+    let (arg_names, arg_types) = args_csv(&ctor.decl.inputs[1..]);
 
     let ctor_payload_unpack = if ctor.decl.inputs.len() > 1 {
         format!(
-            "let CtorPayload {{ {} }} =
+            "let CtorPayload({}) =
                     mantle::reexports::serde_cbor::from_slice(&mantle::backend::input()).unwrap();",
-            ctor_arg_names
+            arg_names
         )
     } else {
         String::new()
     };
+
+    let ctor_stmt = if crate::utils::unpack_syntax_ret(&ctor.decl.output).is_result {
+        format!(
```

Comment:
i don't suppose there's a way to not have this template as a string?

Reply:
Yes unless you want to re-parse the token stream using `syn` just so that you can quote it using `quote` and then re-parse the `proc_macro2::TokenStream` into rustc's `TokenStream`. You'd think otherwise, but quasi-quoting using `quote` is only slightly more ergonomic (repeat patterns and inline quotes) than string formatting.

Maybe someday I'll write a `quote` for rustc TokenStream, but not today.